### ISSUE TITLE
Current version uses a lot of gpu memory

### ISSUE THREAD

**Issue by okihnjo** [EleutherAI/sparsify#79]
I am working with your pretrained SAEs on sentence and token embeddings. I was using the version with commit hash *e6573404b1bee666a99110522547a6232db009ec*, where everything worked fine - total GPU usage was 16GB, 3664 sentences with max size 256 (including padding). The code I was running :
```python
outputs = model(input_ids, attention_mask=attention_mask, output_hidden_states=True)
interesting_layer_BSH = outputs.hidden_states[transformer_layer]
encoded = sae.encode(interesting_layer_BSH)
activations.append(encoded[0].detach().unsqueeze(0))
indices.append(encoded[1].detach().unsqueeze(0))
```

Now, due to cluster reasons, I cloned the newest version of your repository and suddenly, the same code exceeds 40GB of gpu. The error happens in the `sae.encode()` function.
Sorry that I cant be more precise on where this issue exactly occurs but still a thing I wanted to point out.

**Comment by luciaquirke** [comment]
Interesting. Could I have a reproducible version of your code, e.g. with model and sae specified? We have switched to a fused encoder to speed up the backwards pass but I don't think this should increase VRAM usage.

**Comment by okihnjo** [comment]
Sure! This is basically all the necessary code for reproducing the error:
```python
def get_latents(data_loader, model, sae, transformer_layer=11):
    activations = []
    indices = []
    for batch in data_loader: 
        input_ids = batch["input_ids"].to(device)
        attention_mask = batch["attention_mask"].to(device)
        with torch.inference_mode():
            outputs = model(input_ids, attention_mask=attention_mask, output_hidden_states=True)
            interesting_layer_BSH = outputs.hidden_states[transformer_layer]
            encoded = sae.encode(interesting_layer_BSH)
            activations.append(encoded[0].detach().unsqueeze(0))
            indices.append(encoded[1].detach().unsqueeze(0))
    activations_NBSH = torch.vstack(activations)
    indices_NBSH = torch.vstack(indices)
    return activations_NBSH, indices_NBSH

def load_model(model_name):
    bnb_config = BitsAndBytesConfig(load_in_8bit=True)

    model = AutoModelForCausalLM.from_pretrained(
        model_name,
        quantization_config=bnb_config,
    )
    model.config.use_cache = False
    tokenizer = AutoTokenizer.from_pretrained(model_name, padding_side='left')
    tokenizer.pad_token = tokenizer.eos_token
    tokenizer.pad_token_id = tokenizer.eos_token_id
    return model, tokenizer

def load_dataset_features():
    features = Features({
    "input": Value("string"),
    "output": Value("string"),
    "output_concept": Value("string"),
    "concept_genre": Value("string"),
    "category": Value("string"),
    "dataset_category": Value("string"),
    "concept_id": Value("int64"),
    "sae_link": Value("string"),
    "sae_id": Value("int64")
})
    return features


# main function
# --- loading sae ---
sae = Sae.load_from_hub("EleutherAI/sae-llama-3-8b-32x", hookpoint=custom_hookpoint) # custom_hookpoint = 10
# --- loading model and dataset ---
model, tokenizer = load_model(model_name) # model_name = meta--lama/Meta-Llama-3-8B
dataset = load_dataset("pyvene/axbench-concept10", features=load_dataset_features())

activations_stacked_NBTK, indices_stacked_NBTK = get_latents(data_loader, model, sae, transformer_layer=args.transformer_layer + 1)

```
If you need more, let me know. I try to debug it as well this week as you suggested.

**Comment by luciaquirke** [comment]
Your code isn't reproducible out of the box but I spent some time fixing it and after getting it working I am not experiencing your issue - the VRAM usage here is around 12 GBs. 

Since I can't see your working code idk what happened but it's possible that you didn't flatten the batch and sequence dimensions and that caused the TopK to be taken over a smaller dimension than it should have been, such that a larger dimension didn't reduce properly? The fact that SAEs don't accept two batch dimensions is pretty non-obvious.

Code to reproduce:
```
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
from datasets import load_dataset, Features, Value

from sparsify import Sae
from torch.utils.data import DataLoader

device = "cuda"
custom_hookpoint = "layers.10"
transformer_layer = 10


def get_latents(data_loader, model, sae, transformer_layer=11):
    activations = []
    indices = []
    for batch in data_loader:
        input_ids = batch["input_ids"].to(device)
        attention_mask = batch["attention_mask"].to(device)
        with torch.inference_mode():
            outputs = model(
                input_ids, attention_mask=attention_mask, output_hidden_states=True
            )
            interesting_layer_BSH = outputs.hidden_states[transformer_layer]
            encoded = sae.encode(interesting_layer_BSH.flatten(0, 1))
            print("Memory in GBs: ", torch.cuda.memory_allocated() / 1024**3)
            activations.append(encoded[0].detach().unsqueeze(0))
            indices.append(encoded[1].detach().unsqueeze(0))
    activations_NBSH = torch.vstack(activations)
    indices_NBSH = torch.vstack(indices)
    return activations_NBSH, indices_NBSH


def load_model(model_name):
    bnb_config = BitsAndBytesConfig(load_in_8bit=True)

    model = AutoModelForCausalLM.from_pretrained(
        model_name,
        quantization_config=bnb_config,
    )
    model.config.use_cache = False
    tokenizer = AutoTokenizer.from_pretrained(model_name, padding_side="left")
    tokenizer.pad_token = tokenizer.eos_token
    tokenizer.pad_token_id = tokenizer.eos_token_id
    return model, tokenizer


def load_dataset_features():
    features = Features(
        {
            "input": Value("string"),
            "output": Value("string"),
            "output_concept": Value("string"),
            "concept_genre": Value("string"),
            "category": Value("string"),
            "dataset_category": Value("string"),
            "concept_id": Value("int64"),
            "sae_link": Value("string"),
            "sae_id": Value("int64"),
        }
    )
    return features


# main function
# --- loading sae ---
# --- loading model and dataset ---
model, tokenizer = load_model("meta-llama/Meta-Llama-3-8B")
dataset = load_dataset(
    "pyvene/axbench-concept10", features=load_dataset_features(), split="train"
)

print("memory in GBs: ", torch.cuda.memory_allocated() / 1024**3)
sae = Sae.load_from_hub("EleutherAI/sae-llama-3-8b-32x", hookpoint=custom_hookpoint, device=device)
print("memory in GBs: ", torch.cuda.memory_allocated() / 1024**3)

# map input to input_ids and attention_mask from None
dataset = dataset.map(
    lambda x: {
        "input_ids": tokenizer(x["input"], return_tensors="pt").input_ids[0], 
        "attention_mask": tokenizer(x["input"], return_tensors="pt").attention_mask[0]
    }, 
    batched=False
)  # type: ignore
dataset.set_format(type="torch", columns=["input_ids", "attention_mask"])  # type: ignore

data_loader = DataLoader(dataset, batch_size=1, shuffle=False)  # type: ignore

activations_stacked_NBTK, indices_stacked_NBTK = get_latents(
    data_loader, model, sae, transformer_layer=transformer_layer + 1
)
```

**Comment by okihnjo** [comment]
Sorry for the incomplete code, I forgot to add the other bits, I only focused on the `get_latents()`function. 
I am not entirely sure if `interesting_layer_BSH.flatten(0, 1)`is necessary, in the line above I am already accessing a specific layer, so the input to the  `encode()`function should be the same shape as in your readme example
```python
latent_acts = []
    for sae, hidden_state in zip(saes.values(), outputs.hidden_states):
        latent_acts.append(sae.encode(hidden_state))
```
Thanks for having a look though, much appreciated!

**Comment by luciaquirke** [comment]
In the code above the flatten is necessary because the dataloader has multiple items per batch (unlike in the README)

**Comment by okihnjo** [comment]
I always thought the `encode` function expect something like `(num_batches, batch_size, embedding_size)`, because when doing the example in the readme, I get:
```python
# tokenizer = AutoTokenizer.from_pretrained("meta-llama/Meta-Llama-3-8B")
inputs = tokenizer("Hello, world!", return_tensors="pt")
saes = [sae, sae]
with torch.inference_mode():
    model = AutoModelForCausalLM.from_pretrained("meta-llama/Meta-Llama-3-8B")
    outputs = model(**inputs, output_hidden_states=True)
    latent_acts = []
    for sae, hidden_state in zip(saes, outputs.hidden_states):
        print(f"Hidden state shape: {hidden_state.shape}")
        latent_acts.append(sae.encode(hidden_state))
# result
# Hidden state shape: torch.Size([1, 5, 4096])
# Hidden state shape: torch.Size([1, 5, 4096])
```
Looking at the class `FusedEncoder`, the input is indeed defined as 
```python
class FusedEncoder(torch.autograd.Function):
    @staticmethod
    def forward(
        ctx, input,..):
        """
        input:  (N, D)
```
The older version also expects the same shape:
```python
sae_in = x.to(self.dtype)
out = self.encoder(sae_in)
# self.encoder = nn.Linear(d_in, self.num_latents, device=device, dtype=dtype)
return nn.functional.relu(out)
```
I guess I have to rerun my results then, I was always giving the sae an input with 3 dimensions. 
Super weird that I still got somehow results that were kind of okay.
Thank you very much for your time, this was super helpful!

**Comment by luciaquirke** [comment]
Seems like a documentation mistake on our end, oh no 🤦 thanks for the heads up, will fix that ASAP
